File: src/py/ext4-optimizer/lib/ext4listener.py
from subprocess import PIPE, Popen

from lib.utils import clear
from lib.printer import printer
from lib.reader import reader

def ext4listener(language: int, menuable: str = "", echoparts: str = "") -> list[str]:

    COUNT: int = 0; EXTCOUNT: int = 0; MOUNTCOUNT: int = 0
    ABSOLUTEPARTS: str = ""; DIRTYDEVS: list = []
    EXTPARTS: list = []; PARTS: list = []
    UMOUNTS: list[str, str] = []

    ROOT: str = Popen(r"""df -h | sed -ne '/\/$/p' | cut -d " " -f1
                        """, shell=True, stdout=PIPE).stdout.read().decode('utf-8').replace("\n", "")
    VERIFY: str = Popen(r"""find /dev/disk/by-id/ | sort -n | sed 's/^\/dev\/disk\/by-id\///'
                        """, shell=True, stdout=PIPE).stdout.read().decode('utf-8').split("\n")

    for DEVICE in VERIFY:
        DIRTYDEVS.append(Popen(f'readlink "/dev/disk/by-id/{DEVICE}"', shell=True, stdout=PIPE)
                        .stdout.read().decode('utf-8').rstrip().split("\n")[0])

    DIRTYDEVS = list(filter(('').__ne__, DIRTYDEVS))

    for DEV in DIRTYDEVS:
        ABSOLUTEPARTS = Popen(f"""
            echo {DEV} | sed 's/^\.\.\/\.\.\//\/dev\//' | sed '/.*[[:alpha:]]$/d' | sed '/blk[[:digit:]]$/d'""",
            shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
        if ABSOLUTEPARTS != "":
            if ABSOLUTEPARTS != ROOT:
                PARTS.append(Popen(f"echo {DEV} | sed 's/^\.\.\/\.\.\///' | sed '/.*[[:alpha:]]$/d' | sed '/blk[[:digit:]]$/d'",
                    shell=True, stdout=PIPE).stdout.read()
                    .decode('utf-8').rstrip().split("\n")[0]); COUNT += 1

    for PART in PARTS:
        TYPE: str = Popen(f'''lsblk -f /dev/{PART} | sed -ne '2p' | cut -d " " -f2''',
                    shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
        if TYPE == "ext4": EXTCOUNT += 1; EXTPARTS.append(PART)

    if (EXTCOUNT == 0):
        clear(); printer("error", 5, language)
        if menuable == "menu" : input(reader(4)); return
        else: exit(1)

    for PARTITIONSDEF in EXTPARTS:
        MOUNTED: str = Popen(f"lsblk /dev/{PARTITIONSDEF} | sed -ne '/\//p'",
            shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
        if MOUNTED != "": MOUNTCOUNT += 1
        else: UMOUNTS.append(f"/dev/{PARTITIONSDEF}")

    if MOUNTCOUNT == EXTCOUNT :
        clear(); printer("error", 6, language)
        if(menuable == "menu"): input(reader(4, language)); return
        else: exit(1)

    if echoparts == "print": return UMOUNTS

File: src/go/ext4-optimizer/lib/ext4listener.go
package lib

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
)

type OptExtParams struct {
	menuable, echoparts string
}

func Ext4listener(language int, opt OptExtParams) []string {

	EXTCOUNT, MOUNTCOUNT := 0, 0
	ABSOLUTEPARTS, DIRTYDEVS := "", make([]string, 20)
	EXTPARTS, PARTS := make([]string, 40),make([]string, 40)
	UMOUNTS := make([]string, 20)

	SROOT, _ := exec.Command("bash", "-c", "df -h | sed -ne '/\\/$/p' | cut -d\" \" -f1").Output()
	ROOT := strings.Replace(string(SROOT), "\n", "", 1)

	SVERIFY, _ := exec.Command("bash", "-c", "find /dev/disk/by-id/ | sort -n | sed 's/^\\/dev\\/disk\\/by-id\\///'").Output()
	VERIFY := strings.Split(string(SVERIFY), "\n")

	for _, DEVICE := range VERIFY {
		SDIRTDEV, _ := exec.Command("bash", "-c", fmt.Sprintf("readlink \"/dev/disk/by-id/%s\"",DEVICE)).Output()
		DIRTYDEVS = append(DIRTYDEVS, strings.Split(string(SDIRTDEV), "\n")[0])
	}

	DIRTYDEVS = RemoveWhere(DIRTYDEVS,"")

	for _, DEV  := range DIRTYDEVS {
		SABSPARTS, _ := exec.Command("bash", "-c", fmt.Sprintf("echo %s | sed 's/^\\.\\.\\/\\.\\.\\//\\/dev\\//' | sed '/.*[[:alpha:]]$/d' | sed '/blk[[:digit:]]$/d'", DEV)).Output()
		ABSOLUTEPARTS = strings.Replace(string(SABSPARTS), "\n", "", 1)
		if ABSOLUTEPARTS != "" {
			if ABSOLUTEPARTS != ROOT {
				SPARTS, _ := exec.Command("bash", "-c", fmt.Sprintf("echo %s | sed 's/^\\.\\.\\/\\.\\.\\///' | sed '/.*[[:alpha:]]$/d' | sed '/blk[[:digit:]]$/d'", DEV)).Output()
				PARTS = append(PARTS, strings.Split(string(SPARTS), "\n")[0])
			}
		}
	}

	PARTS = RemoveWhere(PARTS,"")

	for _, PART := range PARTS {
		STYPE, _ := exec.Command("bash", "-c", fmt.Sprintf("lsblk -f /dev/%s | sed -ne '2p' | cut -d ' ' -f2", PART)).Output()
		TYPE := strings.Replace(string(STYPE), "\n", "", 1)
		if TYPE == "ext4" {
			EXTCOUNT++
			EXTPARTS = append(EXTPARTS, PART)
		}
	}

	EXTPARTS = RemoveWhere(EXTPARTS,"")

	if EXTCOUNT == 0 {
		Clear(); Printer("error", 5, language)
		if opt.menuable == "menu" {
			fmt.Println(Reader(4, language))
			fmt.Scanln()
		} else {
			os.Exit(1)
		}
	}

	for _, PARTITIONSDEF := range EXTPARTS {
		SMOUNTED, _ := exec.Command("bash", "-c", fmt.Sprintf("lsblk /dev/%s | sed -ne '/\\//p'", PARTITIONSDEF)).Output()
		MOUNTED := strings.Replace(string(SMOUNTED), "\n", "", 1)
		if MOUNTED != "" {
			MOUNTCOUNT++
		} else {
			UMOUNTS = append(UMOUNTS, fmt.Sprintf("/dev/%s", PARTITIONSDEF))
		}
	}

	if MOUNTCOUNT == EXTCOUNT {
		Clear(); Printer("error", 6, language)
		if opt.menuable == "menu" {
			fmt.Println(Reader(4, language))
			fmt.Scanln()
		} else {
			os.Exit(1)
		}
	}

	return UMOUNTS
}
